File: ciber/cAgents/agent.cpp
/*
 * C++ CiberRato agent doing the same as min-01.mus
 */

#include "RobSock.h"

#include <stdio.h>
#include <stdint.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>

/*
 * The simulator has a button that allows to stop the simulation,
 * in which case it not applies orders coming from the agent.
 */
void readSensors()
{
    do { ReadSensors(); } while (GetStartButton() == false);
}

/*
 * This allows synchronize during n cycles,
 * keeping the motors orders sent before.
 */
void wait(unsigned int n)
{
    while (n > 0) { readSensors(); n--; }
}

int main(int argc, char *argv[])
{
    /* connect to simulator */
    const char * name = "Topo Gigio";
    const char * host = "127.0.0.1";
    if (InitRobot(name, 0, host) == -1)
    {
       fprintf(stderr, "Robot \"%s\" fail connecting to simulator @address \"%s\"\n", name, host); 
       exit(1);
    }
    wait(1);

    /* rotate to face North */
    printf("Rotating to face North\n");
    double a = -GetCompassSensor() * M_PI / 180.0;
    double v = (a * 0.5) / 16;
    DriveMotors(-v, v);
    wait(1);

    wait(16-1);

    /* go ahead for 100 cycles */
    printf("Going ahead for %d cycles\n", 100);
    DriveMotors(0.1, 0.1);
    wait(1);

    wait(100-1);

    /* stop robot */
    DriveMotors(0, 0);
    wait(1);
    
    /* signal end of run */
    Finish();
    wait(1);
}

File: ciber/pAgents/agent.py

import sys
import math
import xml.etree.ElementTree as ET

from croblink import *

####################################################################################

CELLROWS = 7
CELLCOLS = 14

####################################################################################

class MyRob(CRobLink):
    def __init__(self, rob_name, rob_id, host):
        CRobLink.__init__(self, rob_name, rob_id, host)

    def readSensors(self):
        while True:
            CRobLink.readSensors(self)
            if self.measures.start: break

    def wait(self, n):
        for i in range(n):
            self.readSensors()

    def run(self):
        if self.status != 0:
            print("Connection refused or error")
            quit()
        self.wait(1)

        # rotate to face North
        print("Rotating to face North")
        a = -self.measures.compass * math.pi / 180.0
        v = (a * 0.5) / 16
        self.driveMotors(-v, v)
        self.wait(1)

        self.wait(15)

        # go ahead for 100 cycles
        print("Going ahead for", 100, "cycles")
        self.driveMotors(0.1, 0.1)
        self.wait(1)

        self.wait(100-1)

        self.driveMotors(0, 0)
        self.wait(1)

        self.finish()
        self.wait(1)


####################################################################################


if __name__ == '__main__':
    name = "Topo gigio"
    host = "127.0.0.1"
    rob = MyRob(name, 0, host)
    
    rob.run()

